packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -1255,6 +1275,15 @@ Hardhat Network's forking functionality only works with blocks from at least spu
             );
           }
 
+          if (
+            this._forkBlockNumber !== undefined &&
+            blockNumber < this._forkBlockNumber
+          ) {
```

Comment:
Addressed this somewhat in https://github.com/nomiclabs/hardhat/pull/1928/commits/e1a246261008fbf582601e9e1a0db68f62022f15 .  I like what it did to the function, and how it approved _some_ of the cases you're talking about.  However, in some areas of the code, we only do `setHardfork()` under certain conditions, so the need to keep those local "is historical?" checks in place anyways, so the impact of this idea turned out to be less than it seemed it would be.

Reply:
I think that we are doing these checks to avoid setting the hardfork to the same value that it already has. If we move the "is historical" condition into the `_selectHardfork` method, we will be setting the HF more often that we strictly need to, but i believe that's innocuos.

Reply:
We're now running with a fresh `Common` each time, addressed in https://github.com/nomiclabs/hardhat/pull/1928/commits/ae26cf55748f9ce3700e1936b8785295d890e94e